src/main/java/com/minecolonies/coremod/entity/ai/citizen/hundertaker/EntityAIWorkUndertaker.java

```diff
@@ -0,0 +1,517 @@
+import com.minecolonies.api.entity.citizen.VisibleCitizenStatus;
+import com.minecolonies.api.tileentities.TileEntityGrave;
+import com.minecolonies.api.util.InventoryUtils;
+import com.minecolonies.api.util.constant.Constants;
+import com.minecolonies.api.util.constant.ToolType;
+import com.minecolonies.api.colony.GraveData;
+import com.minecolonies.coremod.colony.buildings.BuildingMysticalSite;
+import com.minecolonies.coremod.colony.buildings.workerbuildings.BuildingEnchanter;
+import com.minecolonies.coremod.colony.buildings.workerbuildings.BuildingGraveyard;
+import com.minecolonies.coremod.colony.jobs.JobUndertaker;
+import com.minecolonies.coremod.entity.ai.basic.AbstractEntityAIInteract;
+import com.minecolonies.coremod.util.AdvancementUtils;
+import net.minecraft.block.*;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.Direction;
+import net.minecraft.util.Hand;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.Tuple;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TranslationTextComponent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Random;
+
+import static com.minecolonies.api.entity.ai.statemachine.states.AIWorkerState.*;
+import static com.minecolonies.api.research.util.ResearchConstants.RESURRECT_CHANCE;
+import static com.minecolonies.api.util.constant.ToolLevelConstants.TOOL_LEVEL_WOOD_OR_GOLD;
+
+/**
+ * Undertaker AI class.
+ */
+public class EntityAIWorkUndertaker extends AbstractEntityAIInteract<JobUndertaker, BuildingGraveyard>
+{
+    /**
+     * The EXP Earned per dig.
+     */
+    private static final double XP_PER_DIG = 0.5;
+
+    /**
+     * The EXP Earned per wander.
+     */
+    private static final double XP_PER_WANDER = 0.01;
+
+    /**
+     * The weigth of each building level on the resurrection chances.
+     */
+    private static final double RESURRECT_BUILDING_LVL_WEIGHT = 0.005;
+
+    /**
+     * The weigth of each worker level on the resurrection chances.
+     */
+    private static final double RESURRECT_WORKER_MANA_LVL_WEIGHT = 0.0025;
+
+    /**
+     * The max resurrection chance cap [0.0 min -> 1.0 max]
+     */
+    private static final double MAX_RESURRECTION_CHANCE = 0.05;
+
+    /**
+     * The bonus to max resurrection chance cap per max lvl of Mystical Site in the city
+     */
+    private static final double MAX_RESURRECTION_CHANCE_MYSTICAL_LVL_BONUS = 0.01;
+
+    /**
+     * The random variable.
+     */
+    private Random random = new Random();
+
+    /**
+     * A counter to delay some task.
+     */
+    private int effortCounter = 0;
+
+    private static final int EFFORT_EMPTY_GRAVE = 100;
+    private static final int EFFORT_BURY = 400;
+    private static final int EFFORT_RESURRECT = 400;
+
+    /**
+     * Undertaker emptying icon
+     */
+    private final static VisibleCitizenStatus EMPTYING_ICON =
```

Comment:
I'd move the icons somewhere into some constants class, maybe some of the other constants too, to make this easier